op-supervisor/supervisor/backend/db/db.go

```diff
@@ -0,0 +1,494 @@
+	}
+	if (db.lastEntryIdx+1)%searchCheckpointFrequency == 0 {
+		if err := db.writeSearchCheckpoint(block.Number, logIdx, timestamp, block.Hash); err != nil {
+			return fmt.Errorf("failed to write search checkpoint: %w", err)
+		}
+		db.lastEntryContext = postState
+	}
+
+	if err := db.writeInitiatingEvent(postState, log); err != nil {
+		return err
+	}
+	db.lastEntryContext = postState
+	db.updateEntryCountMetric()
+	return nil
+}
+
+// Rewind the database to remove any blocks after headBlockNum
+// The block at headBlockNum itself is not removed.
+func (db *DB) Rewind(headBlockNum uint64) error {
+	db.rwLock.Lock()
+	defer db.rwLock.Unlock()
+	if headBlockNum >= db.lastEntryContext.blockNum {
+		// Nothing to do
+		return nil
+	}
+	// Find the last checkpoint before the block to remove
+	idx, err := db.searchCheckpoint(headBlockNum+1, 0)
+	if errors.Is(err, io.EOF) {
+		// Requested a block prior to the first checkpoint
+		// Delete everything without scanning forward
+		idx = -1
+	} else if err != nil {
+		return fmt.Errorf("failed to find checkpoint prior to block %v: %w", headBlockNum, err)
+	} else {
+		// Scan forward from the checkpoint to find the first entry about a block after headBlockNum
+		i, err := db.newIterator(idx)
+		if err != nil {
+			return fmt.Errorf("failed to create iterator when searching for rewind point: %w", err)
+		}
+		// If we don't find any useful logs after the checkpoint, we should delete the checkpoint itself
+		// So move our delete marker back to include it as a starting point
+		idx--
+		for {
+			blockNum, _, _, err := i.NextLog()
+			if errors.Is(err, io.EOF) {
+				// Reached end of file, we need to keep everything
+				return nil
+			} else if err != nil {
+				return fmt.Errorf("failed to find rewind point: %w", err)
+			}
+			if blockNum > headBlockNum {
+				// Found the first entry we don't need, so stop searching and delete everything after idx
+				break
+			}
+			// Otherwise we need all of the entries the iterator just read
+			idx = i.nextEntryIdx - 1
+		}
+	}
+	// Truncate to contain idx+1 entries, since indices are 0 based, this deletes everything after idx
+	if err := db.data.Truncate((idx + 1) * entrySize); err != nil {
+		return fmt.Errorf("failed to truncate to block %v: %w", headBlockNum, err)
+	}
+	// Update the lastEntryIdx cache and then use db.init() to find the log context for the new latest log entry
+	db.lastEntryIdx = idx
+	if err := db.init(); err != nil {
+		return fmt.Errorf("failed to find new last entry context: %w", err)
+	}
+	return nil
+}
+
+// writeSearchCheckpoint appends search checkpoint and canonical hash entry to the log
+// type 0: "search checkpoint" <type><uint64 block number: 8 bytes><uint32 event index offset: 4 bytes><uint64 timestamp: 8 bytes> = 20 bytes
+// type 1: "canonical hash" <type><parent blockhash truncated: 20 bytes> = 21 bytes
+func (db *DB) writeSearchCheckpoint(blockNum uint64, logIdx uint32, timestamp uint64, blockHash common.Hash) error {
```

Comment:
All the parse functions are pure functions and don't need to be attached as methods to the `*DB`.

All the write functions are almost pure also, except just the last step of writing to the DB, which can probably be decoupled. This would make encode/decode tests easier too.

I think it would be more idiomatic / readable if we define:
-  `type entry [entrySize]byte`
- struct types for each entry variant.

On the struct we can define an `encode()` function that returns an `entry`.

And on the `entry` type we can add functions for each struct, that read it as said struct, with error return argument. Whenever using it with an expected use, you don't have to check the type as caller, but can just call the function, and pass through the `(result, error)` to avoid unnecessary error handling.

Reply:
Yeah I split the encode/decode out to separate methods in https://github.com/ethereum-optimism/optimism/pull/10973/ but wound up not making them struct based. Will have a play with it and see how structs do work out.

The `type entry` is definitely a good idea.

Reply:
Ok, introduced structs for the different entry types and it does clean some stuff up - particularly will be helpful with the executing message support where it wound up parsing the initiatingMessage differently to get the raw diffs rather than resolved block numbers. That now works out much more cleanly.

I haven't added parse methods to the `Entry` type - it's kind of nice that the entry db is a lower level concept that doesn't know anything about the specific entry types.  And I don't think we'd save much error handling since almost every time we parse it's also doing the read and so needs error handling for that anyway. In theory we always know the type when we're parsing and could skip returning an error from the parse code but I kind of like being defensive and checking that the type is what we expect (particularly when reaching search checkpoints from the entry indexes that _should_ always be a search checkpoint).  If we wanted to simplify we could just panic if the entry type is wrong rather than returning an error.